android/jni/com/mapswithme/maps/editor/Editor.cpp

```diff
@@ -351,16 +360,45 @@ Java_com_mapswithme_maps_editor_Editor_nativeCreateMapObject(JNIEnv *, jclass, j
         ("Couldn't create mapobject, wrong coordinates of missing mwm"));
 }
 
+JNIEXPORT void JNICALL
+Java_com_mapswithme_maps_editor_Editor_nativeInitFeatureCategories(JNIEnv * env, jclass clazz)
+{
+  g_featureCategories = make_unique<osm::NewFeatureCategories>(g_framework->NativeFramework()->GetEditorCategories());
+}
+
 JNIEXPORT jobjectArray JNICALL
-Java_com_mapswithme_maps_editor_Editor_nativeGetNewFeatureCategories(JNIEnv * env, jclass clazz)
+Java_com_mapswithme_maps_editor_Editor_nativeGetAllFeatureCategories(JNIEnv * env, jclass clazz, jstring jLang)
 {
-  osm::NewFeatureCategories const & printableTypes = g_framework->NativeFramework()->GetEditorCategories();
-  int const size = printableTypes.m_allSorted.size();
+  InitFeatureCategories();
+
+  string const & lang = jni::ToNativeString(env, jLang);
+  g_featureCategories->AddLanguage(lang);
+  auto const & categories = g_featureCategories->GetAllCategoryNames(lang);
+  int const size = categories.size();
   auto jCategories = env->NewObjectArray(size, g_featureCategoryClazz, 0);
   for (size_t i = 0; i < size; i++)
   {
     // TODO pass used categories section, too
-    jni::TScopedLocalRef jCategory(env, ToJavaFeatureCategory(env, printableTypes.m_allSorted[i]));
+    jni::TScopedLocalRef jCategory(env, ToJavaFeatureCategory(env, categories[i]));
+    env->SetObjectArrayElement(jCategories, i, jCategory.get());
+  }
+
+  return jCategories;
+}
+
+JNIEXPORT jobjectArray JNICALL
+Java_com_mapswithme_maps_editor_Editor_nativeSearchFeatureCategories(JNIEnv * env, jclass clazz, jstring query, jstring jLang)
+{
+  InitFeatureCategories();
```

Comment:
The main problem with InitFeatureCategories() calls is that you need to call it in each method that uses g_featureCategoris.